stellargraph/mapper/graphwave_generator.py

```diff
@@ -0,0 +1,230 @@
+            )
+
+        # Create sparse adjacency matrix:
+        adj = G.to_adjacency_matrix().tocoo()
+
+        # Function to map node IDs to indices for quicker node index lookups
+        self._node_lookup = G._get_index_for_nodes
+
+        degree_mat = diags(np.asarray(adj.sum(1)).ravel())
+        laplacian = degree_mat - adj
+
+        # FIXME(#853): add in option to compute wavelet transform using Chebysev polynomials
+
+        self.scales = np.array(scales).astype(np.float32)
+
+        if num_eigenvecs is None:
+            eigen_vals, eigen_vecs = self._sufficiently_sampled_eigs(
+                laplacian, self.scales.min(), min_delta=min_delta,
+            )
+        else:
+            eigen_vals, eigen_vecs = eigs(laplacian, k=num_eigenvecs, sigma=-1e-3)
+            eigen_vals = np.real(eigen_vals).astype(np.float32)
+
+        self.eigen_vecs = np.real(eigen_vecs).astype(np.float32)
+
+        # the columns of U exp(-scale * eigenvalues) U^T (U = eigenvectors) are used to calculate the node embeddings
+        # (each column corresponds to a node)
+        # to avoid computing a dense NxN matrix when only several eigenvalues are specified
+        # U exp(-scale * eigenvalues) is computed and stored - which is an N x num_eigenvectors matrix
+        # the columns of U exp(-scale * eigenvalues) U^T are then computed on the fly in generator.flow()
+
+        # a list of [U exp(-scale * eigenvalues) for scale in scales]
+        Ues = [self.eigen_vecs * np.exp(-s * eigen_vals) for s in scales]
+        self.Ues = tf.convert_to_tensor(np.stack(Ues, axis=0))
+
+    @staticmethod
+    def _sufficiently_sampled_eigs(laplacian, min_scale, min_delta):
+        """
+        This function calculates increasing numbers of eigenvalues using a binary search until a sufficient number of
+        eigenvalues have been found to ensure accurate results of the GraphWave algorithm.
+
+        Args:
+            laplacian: the un-normalized graph laplacian.
+            min_scale: the smallest wavelet scale used.
+            min_delta: the minimum relative change in the eigenvalue norm. A relative change less than this stops
+                the eigenvalue search.
+        Returns:
+            eigen_vals: the eigenvalues found.
+            eigen_vecs: the eigenvectors found.
+        """
+        max_num_eigs = laplacian.shape[0] - 2
+
+        prev_eig_norm = 0.0
+        eig_max = -1e-3
+        k = min(16, max_num_eigs)
+        eigen_vals = np.array([], dtype=np.float32)
+        eigen_vecs = []
+
+        # use increasing k (doubling per iter) until the filtered eigenvalue l2 norm
+        # increases by less than 10% since the last iter.
+        while True:
+            # if sigma=eigenvalue eigs will throw an error
+            new_eigen_vals, new_eigen_vecs = eigs(laplacian, k=k, sigma=eig_max + 1e-7)
+            new_eigen_vals = np.real(new_eigen_vals).astype(np.float32)
+
+            is_new_eig = new_eigen_vals > eig_max
```

Comment:
what setup gives you the `ValueError`? I haven't come across that yet, although I was getting it today in my new Chebyshev code when the order of the polynomial was too high

Reply:
I had to run it several times (like 10 or more), with `m1 = 10` and `m2 = 11`.

Reply:
found it! turns out `eigs` can return `nan` eigenvectors with valid eigenvalues

Reply:
> nan eigenvectors

Did you mean *vectors* here? If so, I don't think that's the full story, because numpy's warning (and the debugging I did using `warnings.simplefilter("error", RuntimeWarning)` + `import pdb; pdb.pm()`) say that the nan's are in the `new_eigen_vals` array.